File: backend/analyzer.py
import chess
import chess.engine
import chess.pgn
import io
import os
import math
import json
from classifier import classify_move

STOCKFISH_PATH = os.getenv("STOCKFISH_PATH", "/usr/bin/stockfish")

def win_prob(cp: float) -> float:
    # Cap cp to avoid overflow
    cp = max(-10000, min(10000, cp))
    return 50 + 50 * (2 / (1 + math.exp(-0.00368208 * cp)) - 1)

def calculate_accuracy(avg_win_prob_loss: float) -> float:
    if avg_win_prob_loss <= 0: return 100.0
    accuracy = 103.1668 * math.exp(-0.04354 * avg_win_prob_loss) - 3.1669
    return max(0, min(100, accuracy))

async def analyze_game(pgn_str: str, progress_callback=None) -> dict:
    pgn = io.StringIO(pgn_str)
    game = chess.pgn.read_game(pgn)
    if not game:
        raise ValueError("Invalid PGN")

    transport, engine = await chess.engine.popen_uci(STOCKFISH_PATH)
    board = game.board()
    moves_analysis = []
    
    white_cpl = []
    black_cpl = []
    max_cpl = -1
    critical_index = 0

    # Get Opening
    opening = game.headers.get("Opening")
    if not opening:
        eco_url = game.headers.get("ECOUrl")
        if eco_url:
            opening = eco_url.split("/")[-1].replace("-", " ")
        else:
            eco_code = game.headers.get("ECO", "")
            try:
                with open("eco_map.json", "r") as f:
                    eco_map = json.load(f)
                opening = eco_map.get(eco_code, f"{eco_code} Unknown Opening" if eco_code else "Unknown Opening")
            except:
                opening = f"{eco_code} Unknown Opening" if eco_code else "Unknown Opening"

    total_moves = sum(1 for _ in game.mainline_moves())
    
    # Reload game to iterate again
    pgn = io.StringIO(pgn_str)
    game = chess.pgn.read_game(pgn)
    board = game.board()

    for i, move in enumerate(game.mainline_moves()):
        if progress_callback:
            await progress_callback(i, total_moves)

        fen_before = board.fen()
        
        # Analyze current position with multipv
        info_before_list = await engine.analyse(board, chess.engine.Limit(depth=18), multipv=3)
        if not isinstance(info_before_list, list):
            info_before_list = [info_before_list]
            
        info_before = info_before_list[0]
        eval_before_white = info_before["score"].white().score(mate_score=10000)
        eval_before_rel = info_before["score"].relative.score(mate_score=10000)
        best_move = info_before.get("pv", [None])[0]
        
        top_3_moves = []
        for pv_info in info_before_list:
            pv_move = pv_info.get("pv", [None])[0]
            if pv_move:
                san_move = board.san(pv_move)
                score = pv_info["score"].relative.score(mate_score=10000)
                # Convert relative score to white's perspective if needed, but relative is fine for showing alternatives
                top_3_moves.append({"move_san": san_move, "cp": pv_info["score"].white().score(mate_score=10000)})
        
        captured_piece = board.is_capture(move)
        
        board.push(move)
        fen_after = board.fen()
        
        # Eval after move (single PV is fine here since we just need the eval of the position)
        info_after = await engine.analyse(board, chess.engine.Limit(depth=18))
        eval_after_white = info_after["score"].white().score(mate_score=10000)
        eval_after_rel = -info_after["score"].relative.score(mate_score=10000)
        
        cp_loss = max(0, eval_before_rel - eval_after_rel)
        wp_loss = max(0, win_prob(eval_before_rel) - win_prob(eval_after_rel))
        is_best = (move == best_move)
        classification = classify_move(cp_loss, is_best, captured_piece)

        # SAN before push
        board.pop()
        move_san = board.san(move)
        best_move_san = board.san(best_move) if best_move else None
        board.push(move)

        move_data = {
            "move_number": (i // 2) + 1,
            "color": "white" if i % 2 == 0 else "black",
            "move_san": move_san,
            "move_uci": move.uci(),
            "fen_before": fen_before,
            "fen_after": fen_after,
            "eval_before_cp": eval_before_white,
            "eval_after_cp": eval_after_white,
            "best_move_uci": best_move.uci() if best_move else None,
            "best_move_san": best_move_san,
            "top_3_moves": top_3_moves,
            "cp_loss": cp_loss,
            "classification": classification
        }

        moves_analysis.append(move_data)

        if i % 2 == 0: white_cpl.append(wp_loss)
        else: black_cpl.append(wp_loss)

        if cp_loss > max_cpl:
            max_cpl = cp_loss
            critical_index = i

    await engine.quit()

    white_acc = calculate_accuracy(sum(white_cpl)/len(white_cpl)) if white_cpl else 100
    black_acc = calculate_accuracy(sum(black_cpl)/len(black_cpl)) if black_cpl else 100

    return {
        "opening": opening,
        "white_accuracy": white_acc,
        "black_accuracy": black_acc,
        "critical_move_index": critical_index,
        "moves": moves_analysis
    }

File: frontend/lib/analyzer.ts
import { Chess } from 'chess.js';
import { StockfishEngine } from './engine';
import { AnalysisProgress, Move } from '../store/useGameStore';
import { getGameOpening } from './openings';

function winProb(cp: number): number {
  const capped = Math.max(-10000, Math.min(10000, cp));
  return 50 + 50 * (2 / (1 + Math.exp(-0.00368208 * capped)) - 1);
}

function calculateAccuracy(avgWinProbLoss: number): number {
  if (avgWinProbLoss <= 0) return 100.0;
  const accuracy = 103.1668 * Math.exp(-0.04354 * avgWinProbLoss) - 3.1669;
  return Math.max(0, Math.min(100, Math.round(accuracy * 10) / 10));
}

function classifyMove(cpLoss: number, isBest: boolean, isCapture: boolean): string {
  if (isBest || cpLoss <= 10) return 'best';
  if (cpLoss <= 40) return 'good';
  if (cpLoss <= 90) return 'inaccuracy';
  if (cpLoss <= 250) return 'mistake';
  return 'blunder';
}

function uciToSan(fen: string, uci: string | null): string {
  if (!uci || uci.length < 4) return '';
  try {
    const tempChess = new Chess(fen);
    const from = uci.slice(0, 2);
    const to = uci.slice(2, 4);
    const promotion = uci.length > 4 ? uci[4] : undefined;
    const moveObj = tempChess.move({ from, to, promotion });
    return moveObj ? moveObj.san : uci;
  } catch (e) {
    return uci;
  }
}

export async function analyzeGameLocal(pgn: string, onProgress: (progress: AnalysisProgress | string) => void): Promise<any> {
  const chess = new Chess();
  chess.loadPgn(pgn);
  
  const headers = chess.header();
  const whitePlayer = headers['White'] || 'White Player';
  const blackPlayer = headers['Black'] || 'Black Player';
  const whiteElo = headers['WhiteElo'] || '1500';
  const blackElo = headers['BlackElo'] || '1500';
  let openingName = headers['Opening'] || headers['ECO'] || '';
  
  const history = chess.history({ verbose: true });
  const totalMoves = history.length;
  
  const engine = new StockfishEngine();
  const testChess = new Chess();
  
  const movesAnalysis: Move[] = [];
  const whiteWpLosses: number[] = [];
  const blackWpLosses: number[] = [];
  
  onProgress({ status: "Initializing Chess Engine..." });
  
  // Single-pass evaluation: evaluate initial position once
  let currentFen = testChess.fen();
  let currentEngineInfo = await engine.analyzePosition(currentFen, 12);
  
  for (let i = 0; i < totalMoves; i++) {
    const currentLine = history.slice(Math.max(0, i - 4), i + 1).map(m => m.san).join(' ');
    onProgress({
      status: `Analyzing move ${i + 1}/${totalMoves}...`,
      currentMove: i + 1,
      totalMoves,
      currentLine: `...${currentLine}`
    });
    
    const fenBefore = currentFen;
    const move = history[i];
    const isWhiteTurn = i % 2 === 0;
    
    const infoBefore = currentEngineInfo;
    const evalBeforeSide = infoBefore?.evalCp ?? 0;
    const bestMoveUci = infoBefore?.bestMove || '';
    const bestMoveSan = uciToSan(fenBefore, bestMoveUci);
    
    // Play move
    testChess.move(move);
    const fenAfter = testChess.fen();
    
    let evalAfterSide = 0;
    let evalAfterWhite = 0;
    
    if (testChess.isGameOver()) {
      if (testChess.isCheckmate()) {
        // Player who just moved delivered checkmate
        evalAfterSide = 10000;
        evalAfterWhite = isWhiteTurn ? 10000 : -10000;
        currentEngineInfo = { bestMove: null, evalCp: -10000 };
      } else {
        // Stalemate or draw
        evalAfterSide = 0;
        evalAfterWhite = 0;
        currentEngineInfo = { bestMove: null, evalCp: 0 };
      }
    } else {
      currentEngineInfo = await engine.analyzePosition(fenAfter, 12);
      // In fenAfter, it is the opponent's turn.
      // So currentEngineInfo.evalCp is the OPPONENT'S evaluation.
      // Therefore, the evaluation for the player who just moved is -currentEngineInfo.evalCp.
      const oppEval = currentEngineInfo?.evalCp ?? 0;
      evalAfterSide = -oppEval;
      evalAfterWhite = isWhiteTurn ? -oppEval : oppEval;
    }
    
    currentFen = fenAfter;
    
    const moveUci = move.from + move.to + (move.promotion || '');
    const isBest = (bestMoveUci === moveUci);
    
    let cpLoss = isBest ? 0 : Math.max(0, evalBeforeSide - evalAfterSide);
    let wpLoss = isBest ? 0 : Math.max(0, winProb(evalBeforeSide) - winProb(evalAfterSide));
    
    const classification = classifyMove(cpLoss, isBest, 'captured' in move && !!move.captured);
    
    movesAnalysis.push({
      move_number: Math.floor(i / 2) + 1,
      color: isWhiteTurn ? 'white' : 'black',
      move_san: move.san,
      move_uci: moveUci,
      fen_before: fenBefore,
      fen_after: fenAfter,
      eval_before_cp: isWhiteTurn ? evalBeforeSide : -evalBeforeSide,
      eval_after_cp: evalAfterWhite,
      best_move_uci: bestMoveUci,
      best_move_san: bestMoveSan || bestMoveUci,
      cp_loss: cpLoss,
      classification
    });
    
    if (isWhiteTurn) whiteWpLosses.push(wpLoss);
    else blackWpLosses.push(wpLoss);
  }
  
  engine.quit();
  
  // Detect opening if missing from PGN header
  if (!openingName || openingName === 'Custom Opening') {
    const detected = getGameOpening(movesAnalysis);
    openingName = detected && detected.name !== 'Starting Position' ? detected.name : 'Standard Chess Game';
  }
  
  const whiteAcc = calculateAccuracy(whiteWpLosses.reduce((a, b) => a + b, 0) / (whiteWpLosses.length || 1));
  const blackAcc = calculateAccuracy(blackWpLosses.reduce((a, b) => a + b, 0) / (blackWpLosses.length || 1));
  
  onProgress({
    status: "Generating AI suggestions...",
    currentMove: totalMoves,
    totalMoves,
    isComplete: true
  });
  
  // Generate suggestions via Next.js API
  const suggestions = [];
  const criticalMoves = movesAnalysis.filter(m => m.classification === 'blunder' || m.classification === 'mistake').slice(0, 3);
  
  for (const m of criticalMoves) {
    try {
      const res = await fetch('/api/explain', {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({
          fenBefore: m.fen_before,
          playedMoveSan: m.move_san,
          bestMoveSan: m.best_move_san,
          evalCpAfter: m.eval_after_cp,
          cpLoss: m.cp_loss
        })
      });
      const data = await res.json();
      suggestions.push({
        move_index: movesAnalysis.indexOf(m),
        move_san: m.move_san,
        suggestion_text: data.explanation || data.suggestion_text,
        arrow: data.arrow
      });
    } catch (e) {
      console.error(e);
    }
  }

  return {
    opening: openingName,
    white_player: whitePlayer,
    black_player: blackPlayer,
    white_elo: whiteElo,
    black_elo: blackElo,
    white_accuracy: whiteAcc,
    black_accuracy: blackAcc,
    critical_move_index: 0,
    moves: movesAnalysis,
    suggestions: suggestions
  };
}
